### PR TITLE
GG-35449 C++: Java remote filters for CQ  for thick client

### PR DESCRIPTION
**Pull request by isapego**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformCacheEntryEventFilterCpp.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    /** Property to be set from platform. */
+    private long[] longArr;
+
+    /** Injected instance. */
+    @IgniteInstanceResource
+    private Ignite ignite;
+
+    /** {@inheritDoc} */
+    @Override public boolean evaluate(CacheEntryEvent event) throws CacheEntryListenerException {
```
Comment (ademakov) [inline modules/core/src/test/java/org/apache/ignite/platform/PlatformCacheEntryEventFilterCpp.java:87]: I''m totally puzzled about this one. Perhaps a small comment on how this is used would help?

File: modules/platforms/cpp/core/include/ignite/impl/cache/event/cache_entry_event_filter_holder.h [hunk 2]
```diff
@@ -176,6 +178,60 @@ namespace ignite
...
+                    }
+
+                    /**
+                     * Get filter pointer.
+                     *
+                     * @return Filter.
+                     */
+                    virtual CacheEntryEventFilterBase* GetFilter()
+                    {
+                        return 0;
```
Comment (ademakov) [inline modules/platforms/cpp/core/include/ignite/impl/cache/event/cache_entry_event_filter_holder.h:228]: Perhaps add a comment explaining why this is `0`?
And perhaps general info what is the difference between holding `JavaCacheEntryEventFilter` instance vs `CacheEntryEventFilterBase`?

**Review by ademakov — APPROVED** [review]
LGTM

File: modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h [hunk 3]
```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2021 GridGain Systems, Inc. and Contributors.
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h:2]: Agree. Fixed.

File: modules/platforms/cpp/core/include/ignite/impl/writable_object.h [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+             */
+            WritableObject& operator=(const WritableObject& other)
+            {
+                if (this != &other)
+                    val = other.val;
+
+                return *this;
+            }
+
+            virtual ~WritableObject()
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/impl/writable_object.h:108]: Yeah, it's just a part of coding guidelines.

File: modules/platforms/cpp/core/include/ignite/impl/writable_object.h [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+             */
+            WritableObjectInt8Array& operator=(const WritableObjectInt8Array& other)
+            {
+                if (this != &other)
+                    val = other.val;
+
+                return *this;
+            }
+
+            virtual ~WritableObjectInt8Array()
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/impl/writable_object.h:171]: The same answer :)

File: modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            void SetProperty(const std::string& name, const T& value)
+            {
+                std::map<std::string, impl::WritableObjectBase*>::iterator it = properties.find(name);
+                if (it != properties.end())
+                {
+                    delete it->second;
+                    it->second = new impl::WritableObject<T>(value);
+                }
+                else
+                    properties[name] = new impl::WritableObject<T>(value);
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h:112]: You are right, fixed.

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformCacheEntryEventFilterCpp.java (hunk 1 above)
Comment (isapego) [reply modules/core/src/test/java/org/apache/ignite/platform/PlatformCacheEntryEventFilterCpp.java:87]: Well, yeah, this is used in tests to check that object properties passed from C++ to Java layer properly.

File: modules/platforms/cpp/core/include/ignite/cache/query/continuous/continuous_query.h [hunk 7]
```diff
@@ -133,6 +134,35 @@ namespace ignite
...
+                    /**
+                     * Constructor.
+                     *
+                     * @param lsnr Event listener Invoked on the node where
+                     *     continuous query execution has been started.
+                     * @param remoteFilter Java remote filter.
+                     * @param loc Whether query should be executed locally.
+                     */
+                    ContinuousQuery(Reference<event::CacheEntryEventListener<K, V> > lsnr,
+                                    const event::JavaCacheEntryEventFilter& remoteFilter, bool loc) :
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/cache/query/continuous/continuous_query.h:160]: We tend not to use default param values in public API.

File: modules/platforms/cpp/core/include/ignite/impl/cache/event/cache_entry_event_filter_holder.h (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/impl/cache/event/cache_entry_event_filter_holder.h:228]: Added comment. Basically, `JavaCacheEntryEventFilter` is interface that provides us with ability to read remote filter passed to us by Java layer and execute it on specific data. This is needed when we want to use remote filters written in C++ on remote nodes. `JavaCacheEntryEventFilter` on the other hand is just a name of a Java class with the set of parameters, that should be called on Java side and should never be passed to C++ layer.

File: modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ademakov) [reply modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h:2]: Not quite :)

[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/core/include/ignite/impl/platform_java_object_factory_proxy.h:2]: 🤦

**Pull request merged**